# WORK IN PROGRESS

In [1823]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.cm as cm
import seaborn as sns
import warnings

warnings.filterwarnings("ignore")
df_copy.to_csv('data_google_playstore_cleaned.csv')

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
# EXPLORATORY DATA ANALYSIS
## 1. Handling duplicate App
The approach taken here is to create a dataset containing only duplicate apps with differing metadata.
This strategy is intended to simplify operations by isolating the duplicates from the rest of the dataset.

The ultimate goal is to transform the App column into a primary key. Naturally, the process accounts for edge cases where duplicate entries are not errors, but rather distinct records that should be preserved.  It could be the case that two apps with the same name represent different versions of the same app, or even entirely different apps. Only further consideration and assumptions can shed light on this possibility!  

To maintain the integrity of the primary key without discarding these valid variations, some duplicate apps with differing metadata will be renamed by appending a numerical suffix to their name.

***
  
**Disclaimer**  
This analysis might appear excessive, given that the explanation for most discrepancies in metadata across identical app names lies in data entry errors or issues at the source of data collection.  
Additionally, as this is a portfolio project, there is no clearly defined objective beyond conducting a general data analysis. For this reason, a more targeted or goal-specific approach would not have been feasible.

Nevertheless, the analysis serves as a valuable exercise to practice and demonstrate logical reasoning, statistical insight, and coding proficiency — while also ensuring the integrity of the dataset is preserved.

### 1.1 Creation of the new dataset with only rows for duplicate apps 
In this section, alongside the creation of the new dataset, a consistency check is performed to ensure it was generated as intended. A preliminary exploration of the dataset is also conducted—partially addressed earlier during the data cleaning phase when handling duplicates.

To support evaluation and further analysis, it is useful to identify which features vary across duplicate app entries. This includes calculating, for each feature, the percentage of apps that contain differing values within their duplicate group—providing a clear overview of feature-level inconsistency across the dataset

In [1827]:
df_copy2 = df_copy[df_copy['App'].duplicated(keep = False)]

print(f"\n\033[91mNumber of rows in the new dataset - including first instance:\033[0m\n {df_copy2.shape[0]}")
print(f"\n\033[91mNumber of rows in the new dataset - excluding first instance:\033[0m\n {df_copy2['App'].duplicated().sum()}")

print(f"\n\033[91mTotal number of duplicate rows by App in the original dataset - including first instance:\033[0m\n {df_copy['App'].duplicated(keep = False).sum()}")
print(f"\n\033[91mTotal number of duplicate rows by App in the original dataset - excluding first instance:\033[0m\n {df_copy['App'].duplicated().sum()}")

print(f"\n\033[91mTotal number of rows per columns in the original dataset:\033[0m\n {df_copy.shape[0]}")


Number of rows in the new dataset - including first instance:
 1221

Number of rows in the new dataset - excluding first instance:
 698

Total number of duplicate rows by App in the original dataset - including first instance:
 1221

Total number of duplicate rows by App in the original dataset - excluding first instance:
 698

Total number of rows per columns in the original dataset:
 10346


In [1829]:
print("\n\033[91mCount of duplicate apps with different metadata:\033[0m\n")
df_copy[df_copy['App'].duplicated(keep = False)]['App'].value_counts()


Count of duplicate apps with different metadata:



App
ROBLOX                                           9
8 Ball Pool                                      7
Helix Jump                                       6
Zombie Catchers                                  6
Bubble Shooter                                   6
                                                ..
Nike Training Club - Workouts & Fitness Plans    2
Weight Loss Running by Verv                      2
Run with Map My Run                              2
Endomondo - Running & Walking                    2
Maps & GPS Navigation — OsmAnd                   2
Name: count, Length: 523, dtype: int64

In [1831]:
# Exploring the new dataset
df_copy2[df_copy2['App'].duplicated(keep = False)].sort_values(by = 'App').head(10)

App Category  Rating   Reviews        Size  \
5415          365Scores - Live Scores   SPORTS     4.6    666246  25000000.0   
3083          365Scores - Live Scores   SPORTS     4.6    666521  25000000.0   
1755                      8 Ball Pool     GAME     4.5  14200344  52000000.0   
1703                      8 Ball Pool     GAME     4.5  14198602  52000000.0   
1871                      8 Ball Pool     GAME     4.5  14201891  52000000.0   
1844                      8 Ball Pool     GAME     4.5  14200550  52000000.0   
1675                      8 Ball Pool     GAME     4.5  14198297  52000000.0   
1970                      8 Ball Pool     GAME     4.5  14201604  52000000.0   
3953                      8 Ball Pool   SPORTS     4.5  14184910  52000000.0   
559   95Live -SG#1 Live Streaming App   DATING     4.1      4953  15000000.0   

       Installs  Type  Price Content Rating  Genres Last Updated Current Ver  \
5415   10000000  Free    0.0       Everyone  Sports   2018-07-29         559   
3083   10000000  Free    0.0       Everyone  Sports   2018-07-29         559   
1755  100000000  Free    0.0       Everyone  Sports   2018-07-31         400   
1703  100000000  Free    0.0       Everyone  Sports   2018-07-31         400   
1871  100000000  Free    0.0       Everyone  Sports   2018-07-31         400   
1844  100000000  Free    0.0       Everyone  Sports   2018-07-31         400   
1675  100000000  Free    0.0       Everyone  Sports   2018-07-31         400   
1970  100000000  Free    0.0       Everyone  Sports   2018-07-31         400   
3953  100000000  Free    0.0       Everyone  Sports   2018-07-31         400   
559     1000000  Free    0.0           Teen  Dating   2018-08-01         872   

       Android Ver  Date  Month  Year Major  Minor  Patch  
5415    4.1 and up    29      7  2018   559    NaN    NaN  
3083    4.1 and up    29      7  2018   559    NaN    NaN  
1755  4.0.3 and up    31      7  2018   400    NaN    NaN  
1703  4.0.3 and up    31      7  2018   400    NaN    NaN  
1871  4.0.3 and up    31      7  2018   400    NaN    NaN  
1844  4.0.3 and up    31      7  2018   400    NaN    NaN  
1675  4.0.3 and up    31      7  2018   400    NaN    NaN  
1970  4.0.3 and up    31      7  2018   400    NaN    NaN  
3953  4.0.3 and up    31      7  2018   400    NaN    NaN  
559     4.2 and up     1      8  2018   872    NaN    NaN

In [1833]:
# Exploring duplicate apps within a range of occurency of duplicates
app_counts = df_copy['App'].value_counts()
df_copy[df_copy['App'].isin(app_counts[(app_counts > 3) & (app_counts < 6)].index)].sort_values(by='App').head(10)

App  Category  Rating  Reviews  \
2790  AliExpress - Smarter Shopping, Better Living  SHOPPING     4.6  5917485   
4099  AliExpress - Smarter Shopping, Better Living  SHOPPING     4.6  5911055   
2756  AliExpress - Smarter Shopping, Better Living  SHOPPING     4.6  5916569   
2660  AliExpress - Smarter Shopping, Better Living  SHOPPING     4.6  5916606   
1905                           Angry Birds Classic      GAME     4.4  5566889   
3883                           Angry Birds Classic      GAME     4.4  5565856   
1740                           Angry Birds Classic      GAME     4.4  5566805   
1680                           Angry Birds Classic      GAME     4.4  5566669   
1990                           Angry Birds Classic      GAME     4.4  5566908   
4869                               Angry Birds Rio      GAME     4.4  2610328   

            Size   Installs  Type  Price Content Rating    Genres  \
2790         NaN  100000000  Free    0.0           Teen  Shopping   
4099         NaN  100000000  Free    0.0           Teen  Shopping   
2756         NaN  100000000  Free    0.0           Teen  Shopping   
2660         NaN  100000000  Free    0.0           Teen  Shopping   
1905  97000000.0  100000000  Free    0.0       Everyone    Arcade   
3883  97000000.0  100000000  Free    0.0       Everyone    Arcade   
1740  97000000.0  100000000  Free    0.0       Everyone    Arcade   
1680  97000000.0  100000000  Free    0.0       Everyone    Arcade   
1990  97000000.0  100000000  Free    0.0       Everyone    Arcade   
4869  46000000.0  100000000  Free    0.0       Everyone    Arcade   

     Last Updated         Current Ver         Android Ver  Date  Month  Year  \
2790   2018-08-06  Varies with device  Varies with device     6      8  2018   
4099   2018-08-01  Varies with device  Varies with device     1      8  2018   
2756   2018-08-06  Varies with device  Varies with device     6      8  2018   
2660   2018-08-06  Varies with device  Varies with device     6      8  2018   
1905   2018-05-24                 793          4.1 and up    24      5  2018   
3883   2018-05-24                 793          4.1 and up    24      5  2018   
1740   2018-05-24                 793          4.1 and up    24      5  2018   
1680   2018-05-24                 793          4.1 and up    24      5  2018   
1990   2018-05-24                 793          4.1 and up    24      5  2018   
4869   2018-07-03                 269          4.1 and up     3      7  2018   

                   Major  Minor  Patch  
2790  Varies with device    NaN    NaN  
4099  Varies with device    NaN    NaN  
2756  Varies with device    NaN    NaN  
2660  Varies with device    NaN    NaN  
1905                 793    NaN    NaN  
3883                 793    NaN    NaN  
1740                 793    NaN    NaN  
1680                 793    NaN    NaN  
1990                 793    NaN    NaN  
4869                 269    NaN    NaN

In [1836]:
# Individuating features that have distinct records within duplicate apps as max number of distinct records
features=['App', 'Date', 'Month', 'Year', 'Major', 'Minor', 'Patch']

print('\033[91mMax number of distinct values per each app group\033[0m')
for col in df_copy2:
    if col not in features and df_copy2.groupby('App')[col].nunique().max()>1:
        print(f"\n\033[91m{col}\033[0m:", df_copy2.groupby('App')[col].nunique().max())

Max number of distinct values per each app group

Category: 2

Rating: 2

Reviews: 8

Size: 3

Installs: 2

Type: 2

Price: 2

Genres: 3

Last Updated: 3

Current Ver: 3

Android Ver: 2


In [1838]:
print('\n\033[91mPercentage of Apps with Variation per Feature\033[0m')
for col in df_copy2:
    if col not in features:
        sum_var = (df_copy2.groupby('App')[col].nunique() > 1).sum()
        print(f"\n\033[91m{col}:\033[0m {round((sum_var / df_copy2['App'].nunique()) * 100, 2)}%")


Percentage of Apps with Variation per Feature

Category: 16.25%

Rating: 3.25%

Reviews: 98.85%

Size: 2.49%

Installs: 2.87%

Type: 0.38%

Price: 0.38%

Content Rating: 0.0%

Genres: 1.53%

Last Updated: 8.03%

Current Ver: 4.59%

Android Ver: 2.68%


In [1840]:
# Manual check about the discrepancies in records for the duplicate app groups
df_copy2[df_copy2['App'].isin(unique_ver[unique_ver == 2].index)].sort_values(by = 'App').head(40)

App            Category  Rating  \
2310                               Blood Pressure             MEDICAL     4.2   
2513                               Blood Pressure             MEDICAL     NaN   
477                                    Calculator              DATING     2.6   
3242                                   Calculator               TOOLS     4.3   
206                                  Call Blocker            BUSINESS     4.6   
436                                  Call Blocker       COMMUNICATION     4.1   
5044   Cardiac diagnosis (heart rate, arrhythmia)             MEDICAL     4.2   
2269   Cardiac diagnosis (heart rate, arrhythmia)             MEDICAL     4.4   
9641                                   Chess Free                GAME     4.5   
2156                                   Chess Free              FAMILY     4.3   
7329                                  Chrome Beta        PRODUCTIVITY     4.4   
3456                                  Chrome Beta        PRODUCTIVITY     4.4   
820                                        Edmodo           EDUCATION     4.1   
9628                                       Edmodo              FAMILY     4.1   
707                          English Grammar Test           EDUCATION     4.8   
6305                         English Grammar Test              FAMILY     4.6   
10186                            Farm Heroes Saga              FAMILY     4.4   
1880                             Farm Heroes Saga                GAME     4.4   
1733                             Farm Heroes Saga                GAME     4.4   
1695                             Farm Heroes Saga                GAME     4.4   
2011                             Farm Heroes Saga                GAME     4.4   
3292                                   Flashlight               TOOLS     4.7   
3871                                   Flashlight               TOOLS     4.5   
10502                  Fun Kid Racing - Motocross              FAMILY     4.1   
2209                   Fun Kid Racing - Motocross              FAMILY     4.1   
2508                 High Blood Pressure Symptoms             MEDICAL     4.0   
6626                 High Blood Pressure Symptoms             MEDICAL     4.3   
6286                     Moco - Chat, Meet People              DATING     4.2   
489                      Moco - Chat, Meet People              DATING     4.2   
2837                            My Photo Keyboard         PHOTOGRAPHY     4.3   
8446                            My Photo Keyboard         PHOTOGRAPHY     4.1   
1194                                 Panera Bread      FOOD_AND_DRINK     4.2   
8751                                 Panera Bread      FOOD_AND_DRINK     4.2   
2344                               Period Tracker             MEDICAL     4.9   
1357                               Period Tracker  HEALTH_AND_FITNESS     4.5   
2873                                 Photo Editor         PHOTOGRAPHY     4.3   
2933                                 Photo Editor         PHOTOGRAPHY     4.1   
1486                                        Ruler      HOUSE_AND_HOME     4.7   
7570                                        Ruler               TOOLS     4.5   
2238                                 Video Editor              FAMILY     4.1   

       Reviews        Size   Installs  Type  Price Content Rating  \
2310     33033         7.4    5000000  Free   0.00       Everyone   
2513        10         2.4       1000  Free   0.00       Everyone   
477         57         6.2       1000  Paid   6.99       Everyone   
3242     40770         NaN  100000000  Free   0.00       Everyone   
206     188841         3.2    5000000  Free   0.00       Everyone   
436      17529  10000000.0    1000000  Free   0.00       Everyone   
5044      4559         6.4     100000  Free   0.00       Everyone   
2269         8         6.5        100  Paid  12.99       Everyone   
9641   1375988  15000000.0   50000000  Free   0.00       Everyone   
2156     23772  17000000.0    5000

### Insights - new dataset
After a brief inspection of the dataset, it becomes evident that every feature shows multiple distinct values in at least one group of duplicate apps. For instance, Reviews stands out with a remarkable 98.85% variation rate — indicating that nearly every group of duplicate entries differs in this feature.

However, it's important to recognize that not all features represent meaningful differences between duplicate app records. Some variations may be based on dynamic data, technical inconsistencies, or formatting issues rather than actual discrepancies.
This distinction will be addressed at the beginning of the next step.

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
### 1.2 Identifying Discriminant Features
As mentioned, not every feature is relevant in terms of indicating concrete differences among records within the same group of duplicate apps.

The most reasonable approach is to choose, based on domain knowledge, intuition, and common sense. For example, just by quickly looking at the dataset, discrepancies in Category appear to be related to mistakes, as many apps show only one different category across multiple duplicate entries, while other metadata remains identical.  

As a guiding principle, derived from the manual check at the end of the previous section, apps with the same name will be kept as separate records if they differ in key technical features — specifically Current Ver, Price, or Android Ver. These features are assumed to reflect meaningful distinctions between app versions or releases that can be helpful in further analysis regarding versions upgrading.

For this reason, the features kept as discriminants for identifying duplicate apps that might hold distinct and valuable data points are chosen according to the following considerations per feature:
- **Category:** As mentioned, this feature often presents only one differing record among multiple instances of the same duplicate app. In some cases, it's easy to infer that such discrepancies are clearly the result of data collection inaccuracies or incorrect listings from the app developer. One notable case is the app Calculator, which appears under both Tools and Dating.
- **Rating:** Most distinct entries for the same duplicate app show only small variations in rating. These could be due to entry errors at different times or may reflect slight differences across versions.
- **Reviews:** Similar to Rating, variations are frequent but likely tied to time-based updates or versioning.
- **Size:** Discrepancies are most likely due to either data collection errors or different versions of the app. The assumption here is that they are derived from version differences.
- **Installs:** Similar to Size.
- **Genres:** Similar to Category, often inconsistent and potentially non-relevant.
- **Last Updated:** This feature typically varies with the app version, so it is considered a derivative of Current Ver and excluded from independent consideration to reduce redundancy.

***

***Selected Discriminant Features***
- **Price:** This section is emphasized because in several cases, apps with the same name and nearly identical metadata were listed with different prices. This seemed unlikely, even when considering different versions, and therefore suggests either entry errors or version-related distinctions.
- **Current Ver:** This is the core of the distinction between different versions of the same app. It was deemed relevant to preserve different versions, as they might hold valuable information about the app's evolution over time or enable specific analysis by version. Moreover, it was observed that when a group of duplicate apps differed in Current Ver, other metadata such as Reviews, Rating, Size, and Installs often differed as well (and in a few cases, even Category or Price).
- **Android Ver:** Although similar to Current Ver, it was considered relevant to use this feature as an additional filter, since it defines the Android engine compatibility, and thus reflects a structural constraint for each app release.

While it would be impossible to inspect each record individually, some approximations will be performed using imputation methods, particularly for features considered irrelevant.

**Note**
To fully understand the reasons behind duplicate apps with different metadata, a cross-check among various features (e.g., Size vs. Current Ver, Category vs. Price) would be ideal. However, this is beyond the scope of the current portfolio project.
The assumptions stated above should provide a sufficient foundation for handling duplicate apps that may, in fact, represent unique entries.

In [1844]:
# Individuting the amount of apps with different versions
unique_ver = df_copy2.groupby('App')['Current Ver'].nunique()

different_ver = 0
same_ver = 0

for i in unique_ver:
    
    if i > 1:
        different_ver+=1   
    elif i == 1:
        same_ver+=1
        
print(f"\033[91m\nThe amount of apps with different versions is:\033[0m:{different_ver}")
print(f"\033[91m\nThe amount of apps with the same version is:\033[0m: {same_ver}")


The amount of apps with different versions is::24

The amount of apps with the same version is:: 499


In [1846]:
# Individuting the amount of apps with different Current Versions for the same apps
df_copy2.groupby('App')['Current Ver'].nunique().value_counts()

Current Ver
1    499
2     22
3      2
Name: count, dtype: int64

In [1848]:
# Manual check on apps with two or more different Current Versions 
unique_ver = df_copy2.groupby('App')['Current Ver'].nunique()
df_copy2[df_copy2['App'].isin(unique_ver[unique_ver >= 2].index)].sort_values(by='App').head(5)

App Category  Rating  Reviews        Size  Installs  Type  \
2310  Blood Pressure  MEDICAL     4.2    33033         7.4   5000000  Free   
2513  Blood Pressure  MEDICAL     NaN       10         2.4      1000  Free   
3894  Bubble Shooter     GAME     4.5    43576  50000000.0   5000000  Free   
2041  Bubble Shooter   FAMILY     4.4    59843  20000000.0   5000000  Free   
1972  Bubble Shooter     GAME     4.5   148990  46000000.0  10000000  Free   

      Price Content Rating              Genres Last Updated Current Ver  \
2310    0.0       Everyone             Medical   2018-07-24        3273   
2513    0.0       Everyone             Medical   2015-03-14          10   
3894    0.0       Everyone              Arcade   2017-09-13          44   
2041    0.0       Everyone  Puzzle;Brain Games   2018-01-09      233122   
1972    0.0       Everyone              Casual   2018-07-17        1201   

       Android Ver  Date  Month  Year   Major  Minor  Patch  
2310    4.1 and up    24      7  2018    3273    NaN    NaN  
2513    2.3 and up    14      3  2015      10    NaN    NaN  
3894    2.3 and up    13      9  2017      44    NaN    NaN  
2041    2.3 and up     9      1  2018  233122    NaN    NaN  
1972  4.0.3 and up    17      7  2018    1201    NaN    NaN

In [1850]:
# Individuting the amount of apps with different Android Versions for the same apps
df_copy2.groupby('App')['Android Ver'].nunique().value_counts()

Android Ver
1    509
2     14
Name: count, dtype: int64

In [1852]:
# Manual check on apps with two different Android Versions
unique_android = df_copy2.groupby('App')['Android Ver'].nunique()
df_copy2[df_copy2['App'].isin(unique_android[unique_android == 2].index)].sort_values(by='App').head(5)

App Category  Rating  Reviews        Size  Installs  Type  \
2513  Blood Pressure  MEDICAL     NaN       10         2.4      1000  Free   
2310  Blood Pressure  MEDICAL     4.2    33033         7.4   5000000  Free   
2041  Bubble Shooter   FAMILY     4.4    59843  20000000.0   5000000  Free   
3894  Bubble Shooter     GAME     4.5    43576  50000000.0   5000000  Free   
1972  Bubble Shooter     GAME     4.5   148990  46000000.0  10000000  Free   

      Price Content Rating              Genres Last Updated Current Ver  \
2513    0.0       Everyone             Medical   2015-03-14          10   
2310    0.0       Everyone             Medical   2018-07-24        3273   
2041    0.0       Everyone  Puzzle;Brain Games   2018-01-09      233122   
3894    0.0       Everyone              Arcade   2017-09-13          44   
1972    0.0       Everyone              Casual   2018-07-17        1201   

       Android Ver  Date  Month  Year   Major  Minor  Patch  
2513    2.3 and up    14      3  2015      10    NaN    NaN  
2310    4.1 and up    24      7  2018    3273    NaN    NaN  
2041    2.3 and up     9      1  2018  233122    NaN    NaN  
3894    2.3 and up    13      9  2017      44    NaN    NaN  
1972  4.0.3 and up    17      7  2018    1201    NaN    NaN

In [1854]:
# Individuating the amount of apps with different Price for the same apps
df_copy2.groupby('App')['Price'].nunique().value_counts()

Price
1    521
2      2
Name: count, dtype: int64

In [1856]:
unique_price = df_copy2.groupby('App')['Price'].nunique()
df_copy2[df_copy2['App'].isin(unique_price[unique_price == 2].index)].sort_values(by='App')

App Category  Rating  Reviews  \
477                                   Calculator   DATING     2.6       57   
3242                                  Calculator    TOOLS     4.3    40770   
2269  Cardiac diagnosis (heart rate, arrhythmia)  MEDICAL     4.4        8   
5044  Cardiac diagnosis (heart rate, arrhythmia)  MEDICAL     4.2     4559   

      Size   Installs  Type  Price Content Rating   Genres Last Updated  \
477    6.2       1000  Paid   6.99       Everyone   Dating   2017-10-25   
3242   NaN  100000000  Free   0.00       Everyone    Tools   2017-11-21   
2269   6.5        100  Paid  12.99       Everyone  Medical   2018-07-25   
5044   6.4     100000  Free   0.00       Everyone  Medical   2018-07-27   

             Current Ver         Android Ver  Date  Month  Year  \
477                  116          4.0 and up    25     10  2017   
3242  Varies with device  Varies with device    21     11  2017   
2269                   7          3.0 and up    25      7  2018   
5044                 117          3.0 and up    27      7  2018   

                   Major  Minor  Patch  
477                  116    NaN    NaN  
3242  Varies with device    NaN    NaN  
2269                   7    NaN    NaN  
5044                 117    NaN    NaN

In [1858]:
# Total amount of duplicate apps with different price or versions
df_copy2[
    (df_copy2['App'].isin(unique_android[unique_android > 1].index)) |
    (df_copy2['App'].isin(unique_ver[unique_ver > 1].index)) |
    (df_copy2['App'].isin(unique_price[unique_price > 1].index)) 
].shape[0]

59

In [1860]:
# Manual check on duplicate apps with different price or versions
df_copy2[
    (df_copy2['App'].isin(unique_android[unique_android > 1].index)) |
    (df_copy2['App'].isin(unique_ver[unique_ver > 1].index)) |
    (df_copy2['App'].isin(unique_price[unique_price > 1].index)) 
].sort_values(by = 'App').head(5)


App Category  Rating  Reviews        Size  Installs  Type  \
2310  Blood Pressure  MEDICAL     4.2    33033         7.4   5000000  Free   
2513  Blood Pressure  MEDICAL     NaN       10         2.4      1000  Free   
3894  Bubble Shooter     GAME     4.5    43576  50000000.0   5000000  Free   
2041  Bubble Shooter   FAMILY     4.4    59843  20000000.0   5000000  Free   
1972  Bubble Shooter     GAME     4.5   148990  46000000.0  10000000  Free   

      Price Content Rating              Genres Last Updated Current Ver  \
2310    0.0       Everyone             Medical   2018-07-24        3273   
2513    0.0       Everyone             Medical   2015-03-14          10   
3894    0.0       Everyone              Arcade   2017-09-13          44   
2041    0.0       Everyone  Puzzle;Brain Games   2018-01-09      233122   
1972    0.0       Everyone              Casual   2018-07-17        1201   

       Android Ver  Date  Month  Year   Major  Minor  Patch  
2310    4.1 and up    24      7  2018    3273    NaN    NaN  
2513    2.3 and up    14      3  2015      10    NaN    NaN  
3894    2.3 and up    13      9  2017      44    NaN    NaN  
2041    2.3 and up     9      1  2018  233122    NaN    NaN  
1972  4.0.3 and up    17      7  2018    1201    NaN    NaN

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
### 1.3 Renaming apps with different versions or relevant metadata

In [1862]:
# Renaming problematic apps + manual check
problematic_apps = df_copy2[
    (df_copy2['App'].isin(unique_android[unique_android > 1].index)) |
    (df_copy2['App'].isin(unique_ver[unique_ver > 1].index)) |
    (df_copy2['App'].isin(unique_price[unique_price > 1].index)) 
]
app_counter = problematic_apps.groupby('App').cumcount()



df_copy2.loc[problematic_apps.index, 'App'] = np.where(
    app_counter == 0, problematic_apps['App'], 
    problematic_apps['App'] + '(' + app_counter.astype(str) + ')'
)

df_copy2[(df_copy2['App'].str.endswith('(1)')) |
          (df_copy2['App'].str.endswith('(2)')) |
           (df_copy2['App'].str.endswith('(3)'))  ].sort_values(by = 'App')

App        Category  \
2513                                   Blood Pressure(1)         MEDICAL   
1711                                   Bubble Shooter(1)            GAME   
1875                                   Bubble Shooter(2)            GAME   
1972                                   Bubble Shooter(3)            GAME   
3242                                       Calculator(1)           TOOLS   
436                                      Call Blocker(1)   COMMUNICATION   
5044       Cardiac diagnosis (heart rate, arrhythmia)(1)         MEDICAL   
9641                                       Chess Free(1)            GAME   
7329                                      Chrome Beta(1)    PRODUCTIVITY   
9628                                           Edmodo(1)          FAMILY   
6305                             English Grammar Test(1)          FAMILY   
1733                                 Farm Heroes Saga(1)            GAME   
1880                                 Farm Heroes Saga(2)            GAME   
2011                                 Farm Heroes Saga(3)            GAME   
3871                                       Flashlight(1)           TOOLS   
10502                      Fun Kid Racing - Motocross(1)          FAMILY   
6626                     High Blood Pressure Symptoms(1)         MEDICAL   
6286                         Moco - Chat, Meet People(1)          DATING   
8446                                My Photo Keyboard(1)     PHOTOGRAPHY   
8751                                     Panera Bread(1)  FOOD_AND_DRINK   
2344                                   Period Tracker(1)         MEDICAL   
2933                                     Photo Editor(1)     PHOTOGRAPHY   
7570                                            Ruler(1)           TOOLS   
1973                                        Solitaire(1)            GAME   
2024                                        Solitaire(2)          FAMILY   
3712                                     Video Editor(1)   VIDEO_PLAYERS   
5774                                     Video Editor(2)   VIDEO_PLAYERS   
1715                                      Word Search(1)            GAME   
1882                                      Word Search(2)            GAME   
2010                                      Word Search(3)            GAME   
4322   letgo: Buy & Sell Used Stuff, Cars & Real Esta...        SHOPPING   
6194   mySugr: the blood sugar tracker made just for ...         MEDICAL   

       Rating  Reviews        Size   Installs  Type  Price Content Rating  \
2513      NaN       10         2.4       1000  Free    0.0       Everyone   
1711      4.5   148895  46000000.0   10000000  Free    0.0       Everyone   
1875      4.5   148945  46000000.0   10000000  Free    0.0       Everyone   
1972      4.5   148990  46000000.0   10000000  Free    0.0       Everyone   
3242      4.3    40770         NaN  100000000  Free    0.0       Everyone   
436       4.1    17529  10000000.0    1000000  Free    0.0       Everyone   
5044      4.2     4559         6.4     100000  Free    0.0       Everyone   
9641      4.5  1375988  15000000.0   50000000  Free    0.0       Everyone   
7329      4.4   228755         NaN   10000000  Free    0.0       Everyone   
9628      4.1   200214  18000000.0   10000000  Free    0.0       Everyone   
6305      4.6   104676         2.6    5000000  Free    0.0       Everyone   
1733      4.4  7614271  70000000.0  100000000  Free    0.0       Everyone   
1880      4.4  7614415  70000000.0  100000000  Free    0.0       Everyone   
2011      4.4  7614407  70000000.0  100000000  Free    0.0       Everyone   
3871      4.5   162335         4.9   10000000  Free    0.0       Everyone   
10502     4.1    59768         NaN   10000000  Free    0.0       Everyone   
6626      4.3       46         2.0      10000  Free    0.0       Everyone   
6286      4.2   313769         NaN   10000000  Free    0.0     Mature 17+   
8446      4.1   211620  20000000.0   10000000  Free    0.0       Everyone   
8751

In [1866]:
# Finding the number of unique rows (general Python approach)
unique_row = []

for row in df_copy2['App']:
    for num in range(1,10):
        
        if row.endswith('(' + str(num) + ')'):
            unique_row.append(True)
            break
    else:
        unique_row.append(False)
            
df_copy2_filtered = df_copy2[unique_row]
df_copy2_filtered.shape[0]
            
            

35

In [1868]:
# Finding the number of unique rows (Pandas approach)
import re

df_copy2_renamed = df_copy2[df_copy2['App'].str.contains(r'\(\d+\)$')]        # These are the renamed apps
df_copy2_original = df_copy2[~df_copy2['App'].str.contains(r'\(\d+\)$')]     # These are the original apps

print(f"\033[91m\nNumber of rows in duplicate apps dataset:\033[0m {df_copy2.shape[0]}")
print(f"\033[91m\nNumber of renamed apps:\033[0m {df_copy2_renamed.shape[0]}")
print(f"\033[91m\nNumber of original apps:\033[0m {df_copy2_original.shape[0]}")
print('\n(35 + 1186 = 1221)')



Number of rows in duplicate apps dataset: 1221

Number of renamed apps: 35

Number of original apps: 1186

(35 + 1186 = 1221)


In [1870]:
# Step 1: Extract base app names from renamed entries (those ending with (n))
renamed_apps = df_copy2['App'][df_copy2['App'].str.contains(r'\(\d+\)$')]
base_names = renamed_apps.str.replace(r'\(\d+\)$', '', regex=True).unique()

# Step 2: Count how many of these base names appear exactly in the dataset
first_instances = df_copy2[df_copy2['App'].isin(base_names)]

# Step 3: Optional - check how many of these first instances exist
df_copy2_filtered = df_copy2[~df_copy2['App'].str.contains(r'\(\d+\)$')]

print(f"\033[91m\nNumber of original (non-renamed) app instances that also have renamed versions:\033[0m {first_instances.shape[0]}")

print(f"\033[91m\nNumber of duplicated app names (excluding renamed entries):\033[0m {df_copy2_filtered['App'].duplicated(keep=False).sum()}")

print(f"\033[91m\nTotal number of non-renamed app entries:\033[0m {df_copy2_filtered.shape[0]}")



Number of original (non-renamed) app instances that also have renamed versions: 24

Number of duplicated app names (excluding renamed entries): 1162

Total number of non-renamed app entries: 1186


In [1872]:
# df_copy2 = df_copy2.groupby('App').agg({
#     'Reviews': 'mean',
#     'Rating': 'mean',
#     'Size': 'mean',
#     'Category': lambda x: x.mode().iloc[0]
# }).reset_index()

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
### 1.4 Imputation Method for clear duplicate apps
In this section, apps that present clear duplicates — differing only in small values across less relevant features — will be imputed using the following approaches:

- **Reviews:** numerical feature — imputed using the mean
- **Size:** numerical feature — imputed using the mean
- **Rating:** numerical feature — imputed using the mean
- **Category:** categorical feature — imputed using the mode

**NOTE**
Although the most rigorous approach — especially for numerical features — would involve first analyzing the distribution of values within each group of duplicate apps to determine the most appropriate imputation method, this is not feasible in the current context.

The small number of differing values, along with the relatively minor discrepancies observed, allows for confident application of the mean instead of the median. Additionally, performing a group-by-group assessment would be far too time-consuming and beyond the scope of this portfolio project.

In [1876]:
# Fill in the mean and mode values of the related features (e.g. Reviews) across all rows per 'App'
df_copy2['Reviews'] = df_copy2.groupby('App')['Reviews'].transform('mean')
df_copy2['Size'] = df_copy2.groupby('App')['Size'].transform('mean')
df_copy2['Rating'] = df_copy2.groupby('App')['Rating'].transform('mean')
df_copy2['Category'] = df_copy2.groupby('App')['Category'].transform(lambda x: x.mode().iloc[0])


In [1878]:
df_copy2[(
    (df_copy2['App'].duplicated(keep = False)) & (df_copy2['Reviews'].duplicated(keep = False)) &
    df_copy2['Size'].duplicated(keep = False)) & (df_copy2['Rating'].duplicated(keep = False)) &
    (df_copy2['Category'].duplicated(keep = False))].shape[0]

1162

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
### 1.5 Dropping real dentical duplicates for apps and adjourning the original dataset
In this section, the records found where the discrepancies were non importnat will be dropped and the original dataset will be update with only the relevant renamed duplicate apps

In [1882]:
# Count of total duplicate rows in df_copy2 with and without first instance
new_dup = df_copy2.duplicated().sum()
new_dup_tot = df_copy2.duplicated(keep = False).sum()
print(f"\033[91m\nNumber of duplicate rows in df_copy2 (keep='first'):\033[0m {new_dup}")
print(f"\033[91m\nTotal number of duplicate rows in df_copy2 (keep=False):\033[0m {new_dup_tot}")


Number of duplicate rows in df_copy2 (keep='first'): 643

Total number of duplicate rows in df_copy2 (keep=False): 1127


In [1884]:
# Dropping irrelevant duplicate apps
df_copy2 = df_copy2.drop_duplicates(subset = 'App', keep = 'first')
print(f"\033[91m\nShape of df_copy2 after dropping duplicate 'App' entries:\033[0m {df_copy2.shape}")


Shape of df_copy2 after dropping duplicate 'App' entries: (558, 19)


In [1888]:
# Retieving original dataset
original_dtset = df_copy.shape[0]

In [1256]:
# Checking no duplicates are left in the duplicate-only dataset
df_copy2['App'].duplicated(keep = False).sum()

0

In [1894]:
# Dropping all duplicate apps from original dataset for concatenation and check
df_copy = df_copy.drop_duplicates(subset='App', keep=False)

df_copy.duplicated().sum()

0

In [1896]:
# Concatanating togheter the original dataset and the reworked duplicate apps dataset
df_copy = pd.concat([df_copy,df_copy2],axis = 0)
new_dtset = df_copy.shape[0]

In [1898]:
# Comparing original dataset before and after working on duplicate apps
print(f"\033[91m\nShape of the original dataset:\033[0m {original_dtset}")
print(f"\033[91m\nTotal rows of the reworked dataset:\033[0m {new_dtset}")



Shape of the original dataset: 10346

Total rows of the reworked dataset: 9683


In [1900]:
# Checking the reworked original dataset has no more duplicated apps
df_copy.duplicated(subset = 'App').sum()

0

## Insights - duplicate apps
the original dataset went from 10840 entries per 16 columns in in df_copy, dropping those 1181 duplicates.

**The new dataset contains now apps as primary key for easier access!**

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷

## 2. Handling null Rating
### 2.1 Statistical tests for dependancy of missing values
As Mentioned in the data cleaning section (3. IDENTIFYING AND HANDLING MISSING VALUES), the Feature Rating contains a non-negligible amount of null values.  
In this section, said null values will be handled thourgh statistic and imputation methods to improve the integrity of the dataset.

The statistical test that we'll be used here are the **Chi square** to compare Reviews with categorical features and **T-test** or **ANOVA** to compare it to numerical features

1. **Chi square hypotesis:**
- Null hypothesis (H₀): the two features are independent -> no association between the feature and the missingness.
- Alternative hypothesis (H₁): the features are dependent -> there is association between the feature and the missingness.
   
2. **T-test hypothesis:**
- Null hypothesis (H₀): the missingness (in Rating) is independent of the feature → no association
- Alternative hypothesis (H₁): the missingness (in Rating) is dependent on the feature → there is association

For both tests, a p-value > 0.05 indicates failure to reject the null hypothesis, meaning there is association between the features. On the contrary, p-value < 0.05 indicates acceptance of the null hypothesis, or rejection of alternative hypothesis, meaning that there's no association

In [1909]:
df_copy['Rating'].isna().sum()

1459

In [1911]:
df_copy['Rating_missing'] = df_copy['Rating'].isna().astype(int)

In [6]:
# Chi square on categorical feature
from scipy.stats import chi2_contingency

frequency_table = pd.crosstab(df_copy['Category'], df_copy['Rating_missing'])

chi2, p, dof, expected = chi2_contingency(frequency_table)  #Chi squared test

In [1919]:
for col in ['App', 'Category', 'Type', 'Content Rating', 'Genres', 'Current Ver', 'Android Ver']:
    table = pd.crosstab(df_copy[col], df_copy['Rating_missing'])
    chi2, p, dof, expected = chi2_contingency(table)
    print(f"\033[91m\n{col} p-value =\033[0m {p}")



App p-value = 0.4952220533774573

Category p-value = 4.567396120265858e-74

Type p-value = 8.523916356833463e-05

Content Rating p-value = 1.9084401513886045e-10

Genres p-value = 3.639471434845806e-56

Current Ver p-value = 7.3819111792344275e-09

Android Ver p-value = 2.8694041968623634e-24


In [1922]:
# T-test on numerical features
from scipy.stats import ttest_ind

numerical_features = ['Reviews', 'Size', 'Installs', 'Price', 'Date', 'Month', 'Year']
df_copy['Price'] = df_copy['Price'].astype(float)
for col in numerical_features:
    has_rating = df_copy[df_copy['Rating_missing'] == 0][col]
    missing_rating = df_copy[df_copy['Rating_missing'] == 1][col]

    t_stat, p_value = ttest_ind(has_rating, missing_rating, equal_var=False)  # Welch’s t-test
    print(f"\033[91m\n{col} p-value =\033[0m {p_value}")



Reviews p-value = 7.089562184482855e-32

Size p-value = nan

Installs p-value = 2.575861214038849e-46

Price p-value = 0.38860379185581984

Date p-value = 0.004968632033018836

Month p-value = 0.06240540265182251

Year p-value = 0.003886811342580278


## Insights - null Rating
Both statistical tests (Chi-Square for categorical features and T-Test for numerical features) show that missingness in the Rating column is dependent on most features — meaning p-values < 0.05 for the majority of them. Only a few features show p > 0.05, suggesting independence

- **p < 0.05**
Category p-value = 4.56 e-74
Type p-value = 8.52 e-05
Content Rating: p-value = 1.90 e-10
Genres p-value = 3.64 e-56
Current Ver p-value = 2.76 e-05
Android Ver p-value = 2.87 e-24
Reviews p-value = 7.09 e-32
Size p-value = nan
Installs p-value = 2.57 e-46
Date p-value = 0.0049
Year p-value = 0.0038

- **p > 0.05**
App p-value = 0.49
Price p-value = 0.39
Month p-value = 0.062

These results reject the null hypothesis of independence for most features (p < 0.05), meaning there is a significant association between those features and whether a rating is missing.

**Therefore, the missingness in the Rating column is not completely at random (not MCAR)**. Since the missingness is explained by other observed features (but not by Rating itself, which we can't test directly), it satisfies the conditions for being considered:

**Missing At Random (MAR)**

The most appropriate approach is to identify the features most strongly associated with Rating (based on domain knowledge or feature importance) and apply imputation methods based on those features.  
This will help preserve the underlying data structure while reducing bias introduced by the missing value

**NOTE - why two tests?**
Although the initial Chi-Square tests already suggested the missing values in Rating were MAR, the T-Tests were performed as a complementary step to:
- Confirm MAR includes numeric features (like Reviews, Price, Size) — this can inform smarter imputation (e.g. conditional on Reviews).
- Support transparency in EDA/reporting — you're showing your audience all the tested variables.
- Avoid false confidence — what if the missingness didn’t correlate with any categorical features, but strongly correlated with Installs or Price? You’d catch that only via t-test.
- Some ML-based imputation techniques (like KNN or regression imputation) perform better when you know what features the missingness depends on.

In short: T-tests weren’t run to reconfirm MAR, but to provide a complete and feature-inclusive understanding of what influences missingness in the dataset.

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
### 2.2 Null values imputation
In order to perform accurate imputation, it is essential to identify the feature most strongly associated with Rating.
This selection should be guided both by statistical evidence and domain-specific reasoning.

In this case, Category seems the natural choice as it shows a strong association with Rating, supported by an extremely low p-value (4.56e-74) from the Chi-Square test. Additionally, Category has relatively few unique values, which profduces larger group sizes and makes it easier to analyze the distribution of Rating within each group.
Although App would represent the perfect choice given the fact that Rating is logically dependent on such feature, its 1:1 relationship makes it unsuitable for group-based distribution analysis or consistent imputation.

In [1934]:
category_rating_num = df_copy[df_copy['Rating'].isna()]['Category'].nunique()
category_num = df_copy['Category'].nunique()

print(f"\033[91mNumber of unique categories where Rating is nan\033[0m: {category_rating_num}\n\n\033[91mTotal number of unique categories\033[0m: {category_num}")

Number of unique categories where Rating is nan: 32

Total number of unique categories: 33


In [1936]:
category_rating = sorted(df_copy[df_copy['Rating'].isna()]['Category'].unique())
category = sorted(df_copy['Category'].unique())

# category_rating = pd.Series(df_copy3[df_copy3['Rating'].isna()]['Category'].unique()).sort_values
# category = pd.Series(df_copy3['Category'].unique()).sort_values()

print(f"\033[91mCategories where Rating is nan:\033[0m\n\n{category_rating}\n\n\n\033[91mUnique categories:\033[0m\n\n{category}")

Categories where Rating is nan:

['ART_AND_DESIGN', 'AUTO_AND_VEHICLES', 'BEAUTY', 'BOOKS_AND_REFERENCE', 'BUSINESS', 'COMICS', 'COMMUNICATION', 'DATING', 'EDUCATION', 'EVENTS', 'FAMILY', 'FINANCE', 'FOOD_AND_DRINK', 'GAME', 'HEALTH_AND_FITNESS', 'HOUSE_AND_HOME', 'LIBRARIES_AND_DEMO', 'LIFESTYLE', 'MAPS_AND_NAVIGATION', 'MEDICAL', 'NEWS_AND_MAGAZINES', 'PARENTING', 'PERSONALIZATION', 'PHOTOGRAPHY', 'PRODUCTIVITY', 'SHOPPING', 'SOCIAL', 'SPORTS', 'TOOLS', 'TRAVEL_AND_LOCAL', 'VIDEO_PLAYERS', 'WEATHER']


Unique categories:

['ART_AND_DESIGN', 'AUTO_AND_VEHICLES', 'BEAUTY', 'BOOKS_AND_REFERENCE', 'BUSINESS', 'COMICS', 'COMMUNICATION', 'DATING', 'EDUCATION', 'ENTERTAINMENT', 'EVENTS', 'FAMILY', 'FINANCE', 'FOOD_AND_DRINK', 'GAME', 'HEALTH_AND_FITNESS', 'HOUSE_AND_HOME', 'LIBRARIES_AND_DEMO', 'LIFESTYLE', 'MAPS_AND_NAVIGATION', 'MEDICAL', 'NEWS_AND_MAGAZINES', 'PARENTING', 'PERSONALIZATION', 'PHOTOGRAPHY', 'PRODUCTIVITY', 'SHOPPING', 'SOCIAL', 'SPORTS', 'TOOLS', 'TRAVEL_AND_LOCAL', 'VIDEO

In [1940]:
# Assessing which categories have either skewed or normal distribution
skewness = df_copy.groupby('Category')['Rating'].agg(lambda x: x.skew()) # Pandas (vectorized) approach
skewed = skewness[abs(skewness) > 1].index.tolist()
normal = skewness[abs(skewness) <= 1].index.tolist()

# General python approach

# grouped_category = df_copy.groupby('Category')
# skewed = []
# normal = []

# for cat_name, cat_df in grouped_category: 
#     skew_val = cat_df['Rating'].skew()
#     if abs(skew_val) > 1:
#         skewed.append(cat_name)
#     else:
#         normal.append(cat_name)

print(f"\033[91mCategories where Rating distribution is skewed:\033[0m\n\n{skewed}\n\n\n\033[91mCategories where Rating distribution is normal:\033[0m\n\n{normal}")

Categories where Rating distribution is skewed:

['AUTO_AND_VEHICLES', 'BOOKS_AND_REFERENCE', 'BUSINESS', 'COMMUNICATION', 'DATING', 'EVENTS', 'FAMILY', 'FINANCE', 'FOOD_AND_DRINK', 'GAME', 'HEALTH_AND_FITNESS', 'LIFESTYLE', 'MAPS_AND_NAVIGATION', 'MEDICAL', 'NEWS_AND_MAGAZINES', 'PARENTING', 'PERSONALIZATION', 'PHOTOGRAPHY', 'PRODUCTIVITY', 'SHOPPING', 'SOCIAL', 'SPORTS', 'TOOLS', 'TRAVEL_AND_LOCAL', 'VIDEO_PLAYERS']


Categories where Rating distribution is normal:

['ART_AND_DESIGN', 'BEAUTY', 'COMICS', 'EDUCATION', 'ENTERTAINMENT', 'HOUSE_AND_HOME', 'LIBRARIES_AND_DEMO', 'WEATHER']


In [1942]:
df_copy['Rating']  = df_copy['Rating'].fillna(df_copy.groupby('Category')['Rating'].transform('mean'))

# df_copy['Rating'] = np.where(df_copy3['Rating'].isna(), df_copy3.groupby('Category')['Rating'].transform('mean'), df_copy3['Rating'])


    

In [1952]:
# Filling missing values in Rating column based on the Category of each app: median for skewed, mean for normal
# General Python approach
for cat in df_copy['Category'].unique():
    if cat in skewed:
        median_val = df_copy.loc[df_copy['Category'] == cat, 'Rating'].median()
        df_copy.loc[(df_copy['Category'] == cat) & (df_copy['Rating'].isna()), 'Rating'] = median_val
    else:
        mean_val = df_copy.loc[df_copy['Category'] == cat, 'Rating'].mean()
        df_copy.loc[(df_copy['Category'] == cat) & (df_copy['Rating'].isna()), 'Rating'] = mean_val


In [1954]:
# Filling missing values in Rating column based on the Category of each app: median for skewed, mean for normal
# Pandas (vectorized) approach
mask_skewed = (df_copy['Rating'].isna()) & (df_copy['Category'].isin(skewed))
df_copy.loc[mask_skewed, 'Rating'] = df_copy.groupby('Category')['Rating'].transform('median')[mask_skewed]

mask_normal = (df_copy['Rating'].isna()) & (df_copy['Category'].isin(normal))
df_copy.loc[mask_normal, 'Rating'] = df_copy.groupby('Category')['Rating'].transform('mean')[mask_normal]

In [1956]:
# First, create a mask for skewed categories
skewed_mask = df_copy['Category'].isin(skewed)

# Fill skewed categories with median
df_copy.loc[skewed_mask, 'Rating'] = df_copy[skewed_mask].groupby('Category')['Rating'].transform(lambda x: x.fillna(x.median()))

# Fill normal categories with mean
normal_mask = ~skewed_mask
df_copy.loc[normal_mask, 'Rating'] = df_copy[normal_mask].groupby('Category')['Rating'].transform(lambda x: x.fillna(x.mean()))


🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
## 3. Explore Data 

In [ ]:
#here we are going to check which columns are numeric and which are categorical
numeric_feature = [feature for feature in df_copy.columns if df_copy[feature].dtype != 'O']
categorical_feature = [feature for feature in df.columns if df_copy[feature].dtype == 'O']

print('We have {} numerical feature: {}'.format(len(numeric_feature), numeric_feature))
print('We have {} categorical feature: {}'.format(len(categorical_feature), categorical_feature))


The previous line is a list comprehension that generates a list of "features" by iterating through all columns in df_copy.columns. For each column (i.e., feature), it checks the data type in the original dataframe df. If the column's data type is not an object (!= 'O'), meaning it's likely numeric, it includes that feature in the list. In this context, 'O' stands for "object", which usually indicates string data.

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷

In [ ]:
numeric_feature = [feature for feature in df_copy.columns if df_copy[feature].dtype != 'O'

In [ ]:
df_copy['App'].unique()

In [ ]:
for col in categorical_feature:
    print(df[col].value_counts(normalize = True)*100)
    print('-----------------------------------')

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷

In [ ]:
df[df['App']== 'Coloring book moana']

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
## 4. Outliers analysis
**To be completed**

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
## 5. Univariate Analysis of Numeric Features

In [ ]:
plt.figure(figsize = (15,15))
plt.suptitle('Univariate Analysis of Numerical Features', fontsize = 20, fontweight = 'bold', alpha = 0.8, y = 1.)

for i in range(0, len(numeric_feature)):
    plt.subplot(5, 3, i+1)
    sns.kdeplot(x=df_copy[numeric_feature[i]], shade = True, color = 'r')
    plt.xlabel(numeric_feature[i])
    plt.tight_layout()
plt.show()

In [ ]:
df_copy.info()

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷

In [ ]:
df_copy['Rating'].describe()

In [ ]:
df_copy[['Rating']].mode()

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷

In [ ]:
df_copy[['Size']].describe()

In [ ]:
df_copy[['Size']].mode()

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷

In [ ]:
df_copy['Installs'].describe()

In [ ]:
df_copy[['Installs']].mode()

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷

In [ ]:
df_copy[['Year', 'Month']].min()

In [ ]:
df_copy[['Last Updated']].describe()

In [ ]:
df_copy[['Last Updated']].describe()

In [ ]:
df_copy[['Last Updated']].mode()

In [ ]:
df_copy['Last Updated'].value_counts().max()

In [ ]:
mod = df_copy['Last Updated'].mode()[0]
df_copy['Last Updated'].value_counts()[mod]

### Observations
1. **Rating:** right skewed, meaning that, in general, there are more positive ratings than negative with the most ratings being around 4.3. the median being 4.3 confirms the presence of more positive ratings: if 50% of the ratings are higher or equal than 4.3, it means that 50% of the data are between 4.3 and 5 (which clearly is a shorter range than 0 to 4.3).
2. **Reviews:** the count of Reviews doesn't really make much sense since the data in this category is already a count: the number of reviews is hardly the same across multiple apps, there counting how many reviews are the same is, not only not useful, but also conceptually wrong. the best approach here would be to understand what app, genre, maybe compared to number of downloads or to devices, has the biggest number of reviews
3. **Size:** left skewed, meaning that most apps have a smaller size and the most frequent value here (mode since it's a skewed distribution) is 11000000.0 bytes or 11 MB. This is also confirmed by the median = 12 MB, which is closer to the max = 100 MB than to the min =  0.00000095 MB.
4. **Installs:** similarly to Reviews, this is already a count per app, so counting the single value of this feature is a count of count. It would be better instead to analyze these numbers next to the specific app, the category of the app, in a timeframe and/or per type of device
5. **Price:** distribution is right skewed, which tells us that most of the prices for apps is little. The mode = 0 also confirms that most apps are free and just a couple above 350
6. **Last Updated:** the count of this feature tells us that most apps where update in 2018, which is also the year of the series. This is probably consistent with the incremental access to smart phones and, consequently, the usage of apps. The 5-numbers sumary and the shape (left skewed) of the ditribution for this variable seem to be in line with this hypothesis: in fact 50% of the updates took place after 2018-05-03 with the majority of updates on 2018-08-03 (count of mode for 2018-08-03 = 252), which is almost at the end of the series.  

After splitting Last Updated in date, month and year, we can get a better zoom on this variable: the date appear to be somewhat constant with major peaks at the beginning (between the 1st and the 5th) and the end of the month (between the 25th and the 30th); the month(s) with most updates appears to be the 6th, the 7th and the 8th; the years appear to have peaks in 2016, 2017 and 2018, with the last one being extremely higher than the previous two.
We can also see that these last variables are mostly affected by the high number of updates in 2018, so previous years could have different date and month of most updates.

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷

### Univariate Analysis of Categorical features

In [ ]:
df_copy.info()

In [ ]:
df_copy[df_copy['Type'].isnull()]

In [ ]:
(df['Type'].map(type) == str).value_counts()

In [ ]:
(df['Type'].map(type) == str).all()

In [ ]:
df[df['Type'].map(type)!=str]

In [ ]:
df_copy.drop(df_copy.index[9148], inplace = True)

#or

df_copy=df_copy.drop(df_copy.index[9148])

In [ ]:
(df_copy['Type'].map(type) == str).all()

In [ ]:
df_copy['Type'].isnull().sum()

In [ ]:
(df_copy['Type']=='Free').value_counts()

In [ ]:
df_copy['Type'].value_counts().plot(kind = 'bar', color = ['red','blue'])
plt.show()

In [ ]:
df_copy['Content Rating'].value_counts().plot(kind = 'bar', color = 'red')
plt.show()

In [ ]:
#let's merge together the previous two barplots in one plot
series = ['Type','Content Rating']

plt.figure(figsize = (10,10))
for i, col in enumerate(series):
    plt.subplot(2,2,i+1)
    sns.countplot(x = col, data = df_copy, palette = 'Set2')
    plt.xticks(rotation =45)

plt.suptitle('Univariate Analysis of Categorical Features')
plt.show()

In [ ]:
series = ['Type','Content Rating']

plt.figure(figsize = (20,10))

for col in range(0,len(series)):
    plt.subplot(2,2,col+1)
    sns.countplot(x = df[series[col]], palette = 'Set2')
    plt.xlabel(series[col])
    plt.xticks(rotation =45)

plt.suptitle('Univariate Analysis of Categorical Features', fontsize=20, fontweight='bold', alpha=0.8, y=1.)
plt.tight_layout()   
plt.show()

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷
## 6. General Tasks
### What are the most common prices for apps

#### 1.Code version 

In [ ]:
df_copy[['Price']].describe()

In [ ]:
df_copy[['Price']].mode()

In [ ]:
df_copy['Price'].info()

In [ ]:
df_copy['Price'].unique()

In [ ]:
df_copy['Price'].value_counts().sort_index().head(10)

#### 2. Visual version

In [ ]:
bins = range(0, 401, 10)

df_copy['Price_bins'] = pd.cut(df_copy['Price'], bins = bins)
df_copy['Price_bins'].value_counts().sort_index().sort_values(ascending = False).head(10).plot(kind = 'bar', color='red')

plt.xlabel('Price bins = 9.99')
plt.ylabel('Total Price per bin') 
plt.title('Most common ranges for price of apps')
plt.show()

In [ ]:
paid_prices = df_copy[df_copy['Price'] > 0]
categories = paid_prices['Price'].value_counts().sort_index()
categories = categories.sort_values(ascending = False).head(10)
categories = categories.sort_values()
categories.index = categories.index.astype(str) 

plt.figure(figsize=(15, 6))
norm = plt.Normalize(categories.min(), categories.max())
colors = cm.coolwarm(norm(categories)) 
plt.barh(categories.index,categories.values, color = colors)
plt.title('Paid App Prices Only')
plt.xlabel('Price (€)')
plt.ylabel('Number of Apps')
plt.xticks(rotation=90)
plt.tight_layout()
plt.show()


Since the categories are numerical, matplotlib will consider them as continuos. In fact, many values are very close to each other (0.99, 1.0, 10.4) and the consequence is that they will be displayed as a uniqum and not as single categories.  
In order to avoid this problem we need to convert this numbers to categories that Python can recognise as such, this is why we converted them to str with the line:
<span style="color:red">categories.index = categories.index.astype(str).</span> 

This conversion is often needed, especially when:
- Your x-axis values are numeric (like prices or years),
- But you want to treat them as categories (like unique price points),
- And you want a clean bar chart without merging or compression.

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷

#### Most present apps in google playstore

#### 1. Code version

In [ ]:
df_copy[['Category']].value_counts().sort_values(ascending = False).head(10).reset_index(name = 'Count')

In [ ]:
df_copy[['Category']].mode()

In [ ]:
df_copy[['Category']].value_counts().max()

#### 2. Visual version

In [ ]:
top_category = df_copy['Category'].value_counts()

plt.figure(figsize = (15,16))
plt.pie(top_category.values, labels=top_category.index,autopct='%1.1f')
plt.show()

In [ ]:
top_category = pd.DataFrame(df_copy['Category'].value_counts().reset_index())
top_category.columns = ['Category', 'Count'] 

plt.figure(figsize = (15,6))
sns.barplot(x='Category', y = 'Count', data = top_category[:10], palette = 'hls')
plt.xticks(rotation = 45)
plt.title('Top 10 app categories in google playstore')
# plt.xlabel('')
# plt.ylabel('Count of category')
plt.tight_layout()
plt.show()

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷

### Which category has the largest number of installs?

#### 1. Code version

In [ ]:
top_category_df = pd.DataFrame(df_copy.groupby('Category')['Installs'].sum())
top_category_df.sort_values('Installs', ascending = False)

In [ ]:
top_category_df = pd.DataFrame(df_copy.groupby('Category')['Installs'].sum().sort_values(ascending =False).head(5))
top_category_df

In [ ]:
df_copy.groupby('Category')['Installs'].sum().sort_values( ascending = False).to_frame().head(1)

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷

#### 2. Visualization

In [ ]:
top_installs = df_copy.groupby('Category')['Installs'].sum().sort_values(ascending = False).head(20)


plt.figure(figsize=(15,16))
plt.pie(top_installs.values, labels = top_installs.index,autopct = "%1.1f%%",
        labeldistance=1.15,
        pctdistance=0.75)

plt.show()

In [ ]:
installs = df_copy.groupby('Category')['Installs'].sum().sort_values(ascending = False).head(10)



norm = plt.Normalize(installs.min(), installs.max())
colors = cm.coolwarm(norm(installs)) 
plt.figure(figsize = (10,4))
plt.bar(installs.index, installs.values, color=colors)
plt.xticks(rotation=90)

plt.show()

In [ ]:
installs = df_copy.groupby('Category')['Installs'].sum().sort_values(ascending = False).reset_index()
install_sorted = installs.head(10)

palette = sns.color_palette('coolwarm', len(install_sorted))[::-1]
plt.figure(figsize = (10,4))
sns.barplot(x='Installs', y='Category', data = install_sorted, palette = palette)
plt.xticks(rotation=90)
plt.xlabel('Total Instal')
plt.ylabel('Category')
plt.title('Most Downladed Categories')
plt.show()

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷

### What are the top 5 most installed apps in each popular category?

#### 1. Code version

In [ ]:

top_categories = ['GAME', 'COMMUNICATION', 'PRODUCTIVITY', 'SOCIAL' ]

top_apps = {}

for item in top_categories:
    top_apps[item] = df_copy[df_copy['Category']==item].groupby(['App'])['Installs'].sum().sort_values(ascending=False).head(5)




df_top_apps = pd.concat(top_apps).reset_index()
df_top_apps.columns = ['Category', 'App', 'Installs']
df_top_apps

In [ ]:
top_category = ['GAME', 'COMMUNICATION', 'PRODUCTIVITY', 'SOCIAL' ]
dfa = df_copy.groupby(['Category','App'])['Installs'].sum().reset_index()
dfa = dfa.sort_values('Installs', ascending = False)
top_apps=[]

for item in top_category:
    df2 = dfa[dfa['Category'] == item]
    df3 = df2.head(5)
    top_apps.append(df3)

top_apps_df = pd.concat(top_apps, ignore_index=True)
top_apps_df

In [ ]:
df_copy.info()

#### 2 Visualization

In [ ]:
top_category = ['GAME', 'COMMUNICATION', 'PRODUCTIVITY', 'SOCIAL' ]
dfa = df_copy.groupby(['Category','App'])['Installs'].sum().reset_index()
dfa = dfa.sort_values('Installs', ascending = False)
sns.set_context("poster")
sns.set_style("darkgrid")

plt.figure(figsize = (40,30))

for i,item in enumerate(top_category):
    df2 = dfa[dfa['Category'] == item]
    df3 = df2.head(5)
    ax = plt.subplot(4,2,i+1)
    sns.barplot(df3, x='Installs', y='App', palette = 'coolwar_r')
    plt.xlabel('Installs in milions')
    plt.ylabel('')
    ax.set_title(item)

  


plt.tight_layout()
plt.subplots_adjust(hspace= .3)
plt.show()

### How many apps are on google playstore that get 5 ratings?

Let's start by making sure that every app has only one rating

In [ ]:
df[['App','Rating']]

When can make sure that every app has only one rating by making sure that the total number of unique app is the same number of total rows for app.

In [ ]:
df_copy['App'].nunique()

In [ ]:
df_copy['App'].shape[0]

🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷

#### 1. Code Version

In [ ]:
(df_copy['Rating'] == 5).sum()

In [ ]:
df_copy['Rating'].value_counts().sort_index(ascending = False).head(1)


🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷🔷

In [ ]:
my_dict = {x: x**2 for x in range(1,11)}
print(my_dict)

###
#Inverrting keys and values from an existinf dictionary
new_dict = {my_dict[key]:key for key in my_dict} 
print(new_dict)
